File: Module/PlayWays/src/main/java-room-grab/com/module/playways/grab/room/model/SPkRoundInfoModel.java
package com.module.playways.grab.room.model;

import com.alibaba.fastjson.annotation.JSONField;
import com.common.log.MyLog;
import com.module.playways.grab.room.event.GrabSomeOneLightBurstEvent;
import com.module.playways.grab.room.event.GrabSomeOneLightOffEvent;
import com.module.playways.room.song.model.SongModel;
import com.zq.live.proto.Room.EQRoundStatus;
import com.zq.live.proto.Room.EWantSingType;
import com.zq.live.proto.Room.OnlineInfo;
import com.zq.live.proto.Room.QBLightMsg;
import com.zq.live.proto.Room.QCHOInnerRoundInfo;
import com.zq.live.proto.Room.QMLightMsg;
import com.zq.live.proto.Room.QSPKInnerRoundInfo;
import com.zq.live.proto.Room.WantSingInfo;

import org.greenrobot.eventbus.EventBus;

import java.io.Serializable;
import java.util.HashSet;

public class SPkRoundInfoModel implements Serializable {
    public final static String TAG = "SPkRoundInfoModel";

    int userID;
    int singBeginMs;
    int singEndMs;
    private HashSet<BLightInfoModel> bLightInfos = new HashSet<>();//已经爆灯的人, 一唱到底

    private HashSet<MLightInfoModel> mLightInfos = new HashSet<>();//已经灭灯的人, 一唱到底

    private int overReason; // 结束的原因
    private int resultType; // 结果类型
    @JSONField(name = "SPKFinalScore")
    private float score;

    public static SPkRoundInfoModel parse(QSPKInnerRoundInfo roundInfo) {
        SPkRoundInfoModel roundInfoModel = new SPkRoundInfoModel();
        roundInfoModel.setUserID(roundInfo.getUserID());

        roundInfoModel.setSingBeginMs(roundInfo.getSingBeginMs());
        roundInfoModel.setSingEndMs(roundInfo.getSingEndMs());

        roundInfoModel.setOverReason(roundInfo.getOverReason().getValue());
        roundInfoModel.setResultType(roundInfo.getResultType().getValue());

        for (QBLightMsg m : roundInfo.getBLightInfosList()) {
            roundInfoModel.getbLightInfos().add(BLightInfoModel.parse(m));
        }

        for (QMLightMsg m : roundInfo.getMLightInfosList()) {
            roundInfoModel.getMLightInfos().add(MLightInfoModel.parse(m));
        }
        roundInfoModel.setScore(roundInfo.getSPKFinalScore());
        return roundInfoModel;
    }

    public int getUserID() {
        return userID;
    }

    public void setUserID(int userID) {
        this.userID = userID;
    }

    public int getSingBeginMs() {
        return singBeginMs;
    }

    public void setSingBeginMs(int singBeginMs) {
        this.singBeginMs = singBeginMs;
    }

    public int getSingEndMs() {
        return singEndMs;
    }

    public void setSingEndMs(int singEndMs) {
        this.singEndMs = singEndMs;
    }

    public HashSet<BLightInfoModel> getbLightInfos() {
        return bLightInfos;
    }

    public void setbLightInfos(HashSet<BLightInfoModel> bLightInfos) {
        this.bLightInfos = bLightInfos;
    }

    public HashSet<MLightInfoModel> getMLightInfos() {
        return mLightInfos;
    }

    public void setMLightInfos(HashSet<MLightInfoModel> lightInfos) {
        mLightInfos = lightInfos;
    }

    public int getOverReason() {
        return overReason;
    }

    public void setOverReason(int overReason) {
        this.overReason = overReason;
    }

    public int getResultType() {
        return resultType;
    }

    public void setResultType(int resultType) {
        this.resultType = resultType;
    }

    public float getScore() {
        return score;
    }

    public void setScore(float score) {
        this.score = score;
    }

    public void tryUpdateRoundInfoModel(SPkRoundInfoModel roundInfo, boolean notify, GrabRoundInfoModel grabRoundInfoModel) {
        if (roundInfo == null) {
            MyLog.d(TAG, "tryUpdateRoundInfoModel" + " pkRoundInfoModel=" + roundInfo);
            return;
        }
        if (userID == 0) {
            setUserID(roundInfo.getUserID());
        }
        if (roundInfo.getUserID() == userID) {
            this.setSingBeginMs(roundInfo.getSingBeginMs());
            this.setSingEndMs(roundInfo.getSingEndMs());
            for (MLightInfoModel m : roundInfo.getMLightInfos()) {
                addLightOffUid(notify, m, grabRoundInfoModel);
            }
            for (BLightInfoModel m : roundInfo.getbLightInfos()) {
                addLightBurstUid(notify, m, grabRoundInfoModel);
            }

            if (roundInfo.getOverReason() > 0) {
                this.setOverReason(roundInfo.getOverReason());
            }
            if (roundInfo.getResultType() > 0) {
                this.setResultType(roundInfo.getResultType());
            }
            if (roundInfo.getScore() > 0) {
                this.setScore(roundInfo.getScore());
            }
        }
    }

    public boolean addLightOffUid(boolean notify, MLightInfoModel noPassingInfo, GrabRoundInfoModel roundInfoModel) {
        if (!mLightInfos.contains(noPassingInfo)) {
            mLightInfos.add(noPassingInfo);
            if (notify) {
                GrabSomeOneLightOffEvent event = new GrabSomeOneLightOffEvent(noPassingInfo.getUserID(), roundInfoModel);
                EventBus.getDefault().post(event);
            }
            return true;
        }
        return false;
    }

    public boolean addLightBurstUid(boolean notify, BLightInfoModel bLightInfoModel, GrabRoundInfoModel roundInfoModel) {
        if (!bLightInfos.contains(bLightInfoModel)) {
            bLightInfos.add(bLightInfoModel);
            if (notify) {
                GrabSomeOneLightBurstEvent event = new GrabSomeOneLightBurstEvent(bLightInfoModel.getUserID(), roundInfoModel);
                EventBus.getDefault().post(event);
            }
            return true;
        }
        return false;
    }

    @Override
    public String toString() {
        return "SPkRoundInfoModel{" +
                "userID=" + userID +
                ", singBeginMs=" + singBeginMs +
                ", singEndMs=" + singEndMs +
                ", bLightInfos=" + bLightInfos +
                ", mLightInfos=" + mLightInfos +
                ", overReason=" + overReason +
                ", resultType=" + resultType +
                ", score=" + score +
                '}';
    }
}

File: Module/PlayWays/src/main/java-room-grab/com/module/playways/grab/room/model/SPkRoundInfoModel.kt
package com.module.playways.grab.room.model

import com.alibaba.fastjson.annotation.JSONField
import com.common.log.MyLog
import com.module.playways.grab.room.event.GrabSomeOneLightBurstEvent
import com.module.playways.grab.room.event.GrabSomeOneLightOffEvent
import com.module.playways.mic.room.model.MicRoundInfoModel
import com.zq.live.proto.GrabRoom.QBLightMsg
import com.zq.live.proto.GrabRoom.QMLightMsg
import com.zq.live.proto.GrabRoom.QSPKInnerRoundInfo
import com.zq.live.proto.MicRoom.MSPKInnerRoundInfo

import org.greenrobot.eventbus.EventBus

import java.io.Serializable
import java.util.HashSet

class SPkRoundInfoModel : Serializable {
    val TAG = "SPkRoundInfoModel"

    var userID: Int = 0
    var singBeginMs: Int = 0
    var singEndMs: Int = 0
    private var bLightInfos = HashSet<BLightInfoModel>()//已经爆灯的人, 一唱到底

    var mLightInfos = HashSet<MLightInfoModel>()//已经灭灯的人, 一唱到底

    var overReason: Int = 0 // 结束的原因
    var resultType: Int = 0 // 结果类型
    @JSONField(name = "SPKFinalScore")
    var score: Float = 0.toFloat()
    var isWin: Boolean = false

    var meiliTotal: Int = 0   //用来记录魅力值

    fun getbLightInfos(): HashSet<BLightInfoModel> {
        return bLightInfos
    }

    fun setbLightInfos(bLightInfos: HashSet<BLightInfoModel>) {
        this.bLightInfos = bLightInfos
    }

    fun tryUpdateRoundInfoModel(roundInfo: SPkRoundInfoModel?, notify: Boolean, grabRoundInfoModel: GrabRoundInfoModel) {
        if (roundInfo == null) {
            MyLog.d(TAG, "tryUpdateRoundInfoModel pkRoundInfoModel=$roundInfo")
            return
        }
        if (userID == 0) {
            userID = roundInfo.userID
        }
        if (roundInfo.userID == userID) {
            this.singBeginMs = roundInfo.singBeginMs
            this.singEndMs = roundInfo.singEndMs
            for (m in roundInfo.mLightInfos) {
                addLightOffUid(notify, m, grabRoundInfoModel)
            }
            for (m in roundInfo.getbLightInfos()) {
                addLightBurstUid(notify, m, grabRoundInfoModel)
            }

            if (roundInfo.overReason > 0) {
                this.overReason = roundInfo.overReason
            }
            if (roundInfo.resultType > 0) {
                this.resultType = roundInfo.resultType
            }
            if (roundInfo.score > 0) {
                this.score = roundInfo.score
            }
            this.isWin = roundInfo.isWin
        }
    }

    fun tryUpdateRoundInfoModel(roundInfo: SPkRoundInfoModel?) {
        if (roundInfo == null) {
            MyLog.d(TAG, "tryUpdateRoundInfoModel pkRoundInfoModel=$roundInfo")
            return
        }
        if (userID == 0) {
            userID = roundInfo.userID
        }
        if (roundInfo.userID == userID) {
            this.singBeginMs = roundInfo.singBeginMs
            this.singEndMs = roundInfo.singEndMs

            if (roundInfo.overReason > 0) {
                this.overReason = roundInfo.overReason
            }
            if (roundInfo.resultType > 0) {
                this.resultType = roundInfo.resultType
            }
            if (roundInfo.score > 0) {
                this.score = roundInfo.score
            }
            this.isWin = roundInfo.isWin
        }
    }

    fun addLightOffUid(notify: Boolean, noPassingInfo: MLightInfoModel, roundInfoModel: GrabRoundInfoModel): Boolean {
        if (!mLightInfos.contains(noPassingInfo)) {
            mLightInfos.add(noPassingInfo)
            if (notify) {
                val event = GrabSomeOneLightOffEvent(noPassingInfo.getUserID(), roundInfoModel)
                EventBus.getDefault().post(event)
            }
            return true
        }
        return false
    }

    fun addLightBurstUid(notify: Boolean, bLightInfoModel: BLightInfoModel, roundInfoModel: GrabRoundInfoModel): Boolean {
        if (!bLightInfos.contains(bLightInfoModel)) {
            bLightInfos.add(bLightInfoModel)
            if (notify) {
                val event = GrabSomeOneLightBurstEvent(bLightInfoModel.getUserID(), roundInfoModel)
                EventBus.getDefault().post(event)
            }
            return true
        }
        return false
    }

    override fun toString(): String {
        return "SPkRoundInfoModel{" +
                "userID=" + userID +
                ", singBeginMs=" + singBeginMs +
                ", singEndMs=" + singEndMs +
                ", bLightInfos=" + bLightInfos +
                ", mLightInfos=" + mLightInfos +
                ", overReason=" + overReason +
                ", resultType=" + resultType +
                ", score=" + score +
                '}'.toString()
    }

    companion object {

        fun parse(roundInfo: MSPKInnerRoundInfo): SPkRoundInfoModel {
            val roundInfoModel = SPkRoundInfoModel()
            roundInfoModel.userID = roundInfo.userID

            roundInfoModel.singBeginMs = roundInfo.singBeginMs
            roundInfoModel.singEndMs = roundInfo.singEndMs

            roundInfoModel.overReason = roundInfo.overReason.value
            //        roundInfoModel.setResultType(roundInfo.getResultType().getValue());

            roundInfoModel.score = roundInfo.spkFinalScore
            roundInfoModel.isWin = roundInfo.isWin
            return roundInfoModel
        }

        fun parse(roundInfo: QSPKInnerRoundInfo): SPkRoundInfoModel {
            val roundInfoModel = SPkRoundInfoModel()
            roundInfoModel.userID = roundInfo.userID

            roundInfoModel.singBeginMs = roundInfo.singBeginMs
            roundInfoModel.singEndMs = roundInfo.singEndMs

            roundInfoModel.overReason = roundInfo.overReason.value
            roundInfoModel.resultType = roundInfo.resultType.value

            for (m in roundInfo.bLightInfosList) {
                roundInfoModel.getbLightInfos().add(BLightInfoModel.parse(m))
            }

            for (m in roundInfo.mLightInfosList) {
                roundInfoModel.mLightInfos.add(MLightInfoModel.parse(m))
            }
            roundInfoModel.score = roundInfo.spkFinalScore
            roundInfoModel.isWin = roundInfo.isWin
            return roundInfoModel
        }
    }
}
